### PR TITLE
HPCC-29098 Support / in multi-line string at end of line on windows

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/hql/hqlparse.cpp [hunk 1]
```diff
@@ -2146,6 +2154,43 @@ int HqlLex::processStringLiteral(attribute & returnToken, char *CUR_TOKEN_TEXT,
...
+                {
+                    i++;
+                    if ((i+1) < len && text[i+1] == '\n')
+                        i++;
+                    continue;
+                }
+            }
+            else
+            {
+                StringBuffer msg("Can not terminate a string with escape char '\\': ");
```
Comment (shamser) [inline ecl/hql/hqlparse.cpp:2184]: Isn't \ a line joiner rather than escape character?  Also, how do you know that you are at the end of the string here?

**Review by shamser — COMMENTED** [review]
Please can you clarify line 2184-2186.

File: ecl/hql/hqlparse.cpp [hunk 2]
```diff
@@ -2146,6 +2154,43 @@ int HqlLex::processStringLiteral(attribute & returnToken, char *CUR_TOKEN_TEXT,
...
+        char cur = text[i];
+        if (cur == '\\')
+        {
+            if (i+1 < len)
+            {
+                byte next = text[i+1];
+                if (next == '\n')
+                {
+                    i++;
+                    continue;
```
Comment (richardkchapman) [inline ecl/hql/hqlparse.cpp:2172]: Does this continue do anything?

File: ecl/hql/hqlparse.cpp [hunk 3]
```diff
@@ -2146,6 +2154,43 @@ int HqlLex::processStringLiteral(attribute & returnToken, char *CUR_TOKEN_TEXT,
...
+                }
+                if (next == '\r')
+                {
+                    i++;
+                    if ((i+1) < len && text[i+1] == '\n')
+                        i++;
+                    continue;
```
Comment (richardkchapman) [inline ecl/hql/hqlparse.cpp:2179]: Or this one?

File: ecl/hql/hqlparse.cpp [hunk 4]
```diff
@@ -1997,6 +1997,14 @@ int HqlLex::processStringLiteral(attribute & returnToken, char *CUR_TOKEN_TEXT,
                 finger++;
                 continue;  // A \ at end of line in a multiline constant means remove the end-of-line
             }
+            else if (next == '\r')
```
Comment (richardkchapman) [inline ecl/hql/hqlparse.cpp:2000]: Feels like something bigger is wrong if a multiline string literal from a Windows source file has a different contents than from a non-windows one. Isn't a more correct fix to have translate \r\n to \n earlier?

**Review by richardkchapman — CHANGES_REQUESTED** [review]
This feels like the wrong fix to me - suggests the "dossiness" of the source file has an impact on its meaning which doesn't feel desirable. Perhaps something to discuss?

**Comment by ghalliday** [comment]
@richardkchapman what do you think about these changes?

**Review by richardkchapman — APPROVED** [review]
Looks good - one minor quibble is that the function "stripSlashNewline" now does a little more than just what its name suggests.

**Review by shamser — APPROVED** [review]

**Pull request merged**